subsys/bluetooth/mesh/adv_legacy.c

```diff
@@ -243,7 +248,11 @@ int bt_mesh_adv_disable(void)
 {
 	int err;
 
-	enabled = false;
+	/* k_thread_join will sleep forever if BT_MESH_SUSPENDED flag is not set. The advertiser
+	 * thread will exit once the flag is set. The flag is set by the higher layer function. Here
+	 * we need to check that the flag is dropped and ensure that the thread is stopped.
+	 */
+	__ASSERT_NO_MSG(!ADV_ENABLED);
```

Comment:
This 'ensure' works only when asserts are enabled, otherwise this can cause locking, better to return error code to let higher layer handle it.